pkg/controllers/dynakube/otelc/secret/reconciler.go

```diff
@@ -0,0 +1,160 @@
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Reconciler struct {
+	client    client.Client
+	apiReader client.Reader
+	dk        *dynakube.DynaKube
+}
+
+type ReconcilerBuilder func(client client.Client, apiReader client.Reader, dk *dynakube.DynaKube) *Reconciler
+
+func NewReconciler(client client.Client, apiReader client.Reader, dk *dynakube.DynaKube) *Reconciler {
+	return &Reconciler{
+		client:    client,
+		dk:        dk,
+		apiReader: apiReader,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context) error {
+	if r.dk.TelemetryService().IsEnabled() {
+		return r.ensureOpenSignalAPISecret(ctx)
+	}
+
+	return r.removeOpenSignalAPISecret(ctx)
+}
+
+func (r *Reconciler) ensureOpenSignalAPISecret(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.apiReader, log)
+	_, err := query.Get(ctx, types.NamespacedName{Name: consts.TelemetryApiCredentialsSecretName, Namespace: r.dk.Namespace})
+
+	if err != nil && k8serrors.IsNotFound(err) {
+		log.Info("creating new secret for telemetry api credentials")
+
+		secretConfig, err := r.generateTelemetryApiCredentialsSecret(ctx, consts.TelemetryApiCredentialsSecretName)
+
+		if err != nil {
+			conditions.SetSecretGenFailed(r.dk.Conditions(), secretConditionType, err)
+
+			return err
+		}
+
+		_, err = hasher.GenerateHash(secretConfig.Data)
+		if err != nil {
+			conditions.SetSecretGenFailed(r.dk.Conditions(), secretConditionType, err)
+
+			return err
+		}
+
+		err = query.Create(ctx, secretConfig)
+		if err != nil {
+			log.Info("could not create secret for telemetry api credentials", "name", secretConfig.Name)
+			conditions.SetKubeApiError(r.dk.Conditions(), secretConditionType, err)
+
+			return err
+		}
+
+		conditions.SetSecretCreated(r.dk.Conditions(), secretConditionType, consts.TelemetryApiCredentialsSecretName)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) getApiToken(ctx context.Context) ([]byte, error) {
+	tokenReader := token.NewReader(r.apiReader, r.dk)
+
+	tokens, err := tokenReader.ReadTokens(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "'%s:%s' secret is missing or invalid", r.dk.Namespace, r.dk.Tokens())
+	}
+
+	apiToken, hasApiToken := tokens[dtclient.ApiToken]
+	if !hasApiToken {
+		return nil, errors.New(fmt.Sprintf("'%s' token is missing in '%s:%s' secret", dtclient.ApiToken, r.dk.Namespace, r.dk.Tokens()))
+	}
+
+	return []byte(apiToken.Value), nil
+}
+
+func (r *Reconciler) getDtEndpoint() ([]byte, error) {
+	tenantUUID, err := r.dk.TenantUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	if r.dk.ActiveGate().IsApiEnabled() {
+		return []byte(fmt.Sprintf("https://%s-activegate.dynatrace.svc/e/%s/api/v2/otlp", r.dk.Name, tenantUUID)), nil
+	}
+
+	return []byte(fmt.Sprintf("https://%s.dev.dynatracelabs.com/api/v2/otlp", tenantUUID)), nil
```

Comment:
let's make sure the endpoint is correct

Reply:
this endpoint from requirements from the story :) cc @StefanHauth

Reply:
This should facilitate the API URL configured in the DynaKube, because the public endpoint differs between dev and prod environments. The cannoit be hardcoded here.

Reply:
fixed in 3a27c5763